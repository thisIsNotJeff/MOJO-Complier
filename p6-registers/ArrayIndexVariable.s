	.data
	.balign 8
ArrayIndexVariable.x:
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.text
ArrayIndexVariable.Get:
	pushq %rbp
	movq %rsp,%rbp
	subq $16,%rsp
	movq %rbx, -8(%rbp)
	movq %rdi,%rax
L.3:
	leaq ArrayIndexVariable.x(%rip),%rbx
#	movq %rbx,%rbx
#	movq %rax,%rax
	shlq $3,%rax
	addq %rax,%rbx
	movq 0(%rbx),%rax
#	movq %rax,%rax
L.0:
#	returnSink
	movq -8(%rbp),%rbx
	addq $16,%rsp
	popq %rbp
	ret
	.text
ArrayIndexVariable:
	pushq %rbp
	movq %rsp,%rbp
#	returnSink
	popq %rbp
	ret
	.globl main
	.text
main:
	pushq %rbp
	movq %rsp,%rbp
L.4:
	call ArrayIndexVariable
#	returnSink
	popq %rbp
	ret
	.data
	.balign 8
L.1:	.asciz	"Attempt to use a null pointer"
	.text
badPtr:
	pushq %rbp
	movq %rsp,%rbp
L.5:
	leaq L.1(%rip),%rax
	movq %rax,%rdi
	call puts
	movabsq $1,%rax
	movq %rax,%rdi
	call exit
#	returnSink
	popq %rbp
	ret
	.data
	.balign 8
L.2:	.asciz	"Subscript out of bounds"
	.text
badSub:
	pushq %rbp
	movq %rsp,%rbp
L.6:
	leaq L.2(%rip),%rax
	movq %rax,%rdi
	call puts
	movabsq $1,%rax
	movq %rax,%rdi
	call exit
#	returnSink
	popq %rbp
	ret
