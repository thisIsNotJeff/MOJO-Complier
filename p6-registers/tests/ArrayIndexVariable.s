	.data
	.balign 8
ArrayIndexVariable.x:
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.text
ArrayIndexVariable.Get:
	pushq %rbp
	movq %rsp,%rbp
#	movq %rdi,%rdi
L.3:
	leaq ArrayIndexVariable.x(%rip),%rax
#	movq %rax,%rax
#	movq %rdi,%rdi
	shlq $3,%rdi
	addq %rdi,%rax
	movq 0(%rax),%rax
#	movq %rax,%rax
L.0:
#	returnSink
	popq %rbp
	ret
	.text
ArrayIndexVariable:
	pushq %rbp
	movq %rsp,%rbp
#	returnSink
	popq %rbp
	ret
	.globl main
	.text
main:
	pushq %rbp
	movq %rsp,%rbp
L.4:
	call ArrayIndexVariable
#	returnSink
	popq %rbp
	ret
	.data
	.balign 8
L.1:	.asciz	"Attempt to use a null pointer"
	.text
badPtr:
	pushq %rbp
	movq %rsp,%rbp
L.5:
	leaq L.1(%rip),%rdi
#	movq %rdi,%rdi
	call puts
	movabsq $1,%rdi
#	movq %rdi,%rdi
	call exit
#	returnSink
	popq %rbp
	ret
	.data
	.balign 8
L.2:	.asciz	"Subscript out of bounds"
	.text
badSub:
	pushq %rbp
	movq %rsp,%rbp
L.6:
	leaq L.2(%rip),%rdi
#	movq %rdi,%rdi
	call puts
	movabsq $1,%rdi
#	movq %rdi,%rdi
	call exit
#	returnSink
	popq %rbp
	ret
