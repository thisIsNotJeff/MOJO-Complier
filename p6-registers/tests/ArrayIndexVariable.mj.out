	.data
	.balign 8
ArrayIndexVariable.x:
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
	.quad 0
PROCEDURE ArrayIndexVariable.Get
# Before canonicalization (trees):
MOVE(
 TEMP %rax,
 MEM(
  ADD(
   NAME ArrayIndexVariable.x,
   MUL(
    TEMP ArrayIndexVariable.Get.n,
    CONST 8)), 
  CONST 0, 8)),
JUMP(
 NAME L.0),
LABEL L.0
# After canonicalization (trees):
MOVE(
 TEMP %rax,
 MEM(
  ADD(
   NAME ArrayIndexVariable.x,
   MUL(
    TEMP ArrayIndexVariable.Get.n,
    CONST 8)), 
  CONST 0, 8))
JUMP(
 NAME L.0)
LABEL L.0
# Basic Blocks:
LABEL L.3
MOVE(
 TEMP %rax,
 MEM(
  ADD(
   NAME ArrayIndexVariable.x,
   MUL(
    TEMP ArrayIndexVariable.Get.n,
    CONST 8)), 
  CONST 0, 8))
JUMP(
 NAME L.0)
LABEL L.0
# Trace Scheduled:
LABEL L.3
MOVE(
 TEMP %rax,
 MEM(
  ADD(
   NAME ArrayIndexVariable.x,
   MUL(
    TEMP ArrayIndexVariable.Get.n,
    CONST 8)), 
  CONST 0, 8))
LABEL L.0
# With procedure entry/exit:
MOVE(
 TEMP ArrayIndexVariable.Get.n,
 TEMP %rdi)
LABEL L.3
MOVE(
 TEMP %rax,
 MEM(
  ADD(
   NAME ArrayIndexVariable.x,
   MUL(
    TEMP ArrayIndexVariable.Get.n,
    CONST 8)), 
  CONST 0, 8))
LABEL L.0
# Instructions:
	movq %rdi,ArrayIndexVariable.Get.n	# ArrayIndexVariable.Get.n := %rdi 
L.3:	# <- 
	leaq ArrayIndexVariable.x(%rip),t.1	# t.1 <- 
	movq t.1,t.0	# t.0 := t.1 
	movq ArrayIndexVariable.Get.n,t.2	# t.2 := ArrayIndexVariable.Get.n 
	shlq $3,t.2	# t.2 <- t.2 
	addq t.2,t.0	# t.0 <- t.2 t.0 
	movq 0(t.0),t.3	# t.3 <- t.0 
	movq t.3,%rax	# %rax := t.3 
L.0:	# <- 
# Control Flow Graph:
0: ArrayIndexVariable.Get.n <- %rdi ; goto 1 
1: t.1 t.0 t.2 t.3 %rax <- ArrayIndexVariable.Get.n ; goto 2 
2: <- %rsp %rbp %rax ; goto 
# Interference Graph:
%rax: 
%rbx: 
%rcx: 
%rdx: 
%rsp: 
%rbp: 
%rsi: 
%rdi: 
%r8: 
%r9: 
%r10: 
%r11: 
%r12: 
%r13: 
%r14: 
%r15: 
ArrayIndexVariable.Get.n: %rsp %rbp t.0 t.1 
t.1: %rsp %rbp ArrayIndexVariable.Get.n 
t.0: %rsp %rbp t.2 ArrayIndexVariable.Get.n 
t.2: %rsp %rbp t.0 
t.3: %rsp %rbp 
t.0 <= t.1
t.2 <= ArrayIndexVariable.Get.n
%rax <= t.3
ArrayIndexVariable.Get.n <= %rdi
# Register Allocation:
%rax->%rax
%rbx->%rbx
%rcx->%rcx
%rdx->%rdx
%rsp->%rsp
%rbp->%rbp
%rsi->%rsi
%rdi->%rdi
%r8->%r8
%r9->%r9
%r10->%r10
%r11->%r11
%r12->%r12
%r13->%r13
%r14->%r14
%r15->%r15
ArrayIndexVariable.Get.n->%rdi
t.1->%rax
t.0->%rax
t.2->%rdi
t.3->%rax
# Assembly code:
	.text
ArrayIndexVariable.Get:
	pushq %rbp
	movq %rsp,%rbp
#	movq %rdi,%rdi
L.3:
	leaq ArrayIndexVariable.x(%rip),%rax
#	movq %rax,%rax
#	movq %rdi,%rdi
	shlq $3,%rdi
	addq %rdi,%rax
	movq 0(%rax),%rax
#	movq %rax,%rax
L.0:
#	returnSink
	popq %rbp
	ret
END ArrayIndexVariable.Get
PROCEDURE ArrayIndexVariable
# With procedure entry/exit:
# Instructions:
# Control Flow Graph:
0: <- %rsp %rbp ; goto 
# Interference Graph:
%rax: 
%rbx: 
%rcx: 
%rdx: 
%rsp: 
%rbp: 
%rsi: 
%rdi: 
%r8: 
%r9: 
%r10: 
%r11: 
%r12: 
%r13: 
%r14: 
%r15: 
# Register Allocation:
%rax->%rax
%rbx->%rbx
%rcx->%rcx
%rdx->%rdx
%rsp->%rsp
%rbp->%rbp
%rsi->%rsi
%rdi->%rdi
%r8->%r8
%r9->%r9
%r10->%r10
%r11->%r11
%r12->%r12
%r13->%r13
%r14->%r14
%r15->%r15
# Assembly code:
	.text
ArrayIndexVariable:
	pushq %rbp
	movq %rsp,%rbp
#	returnSink
	popq %rbp
	ret
END ArrayIndexVariable
PROCEDURE main
# Before canonicalization (trees):
EXP(
 CALL(
  NAME ArrayIndexVariable,
  CONST 0))
# After canonicalization (trees):
EXP(
 CALL(
  NAME ArrayIndexVariable,
  CONST 0))
# Basic Blocks:
LABEL L.4
EXP(
 CALL(
  NAME ArrayIndexVariable,
  CONST 0))
# Trace Scheduled:
LABEL L.4
EXP(
 CALL(
  NAME ArrayIndexVariable,
  CONST 0))
# With procedure entry/exit:
LABEL L.4
EXP(
 CALL(
  NAME ArrayIndexVariable,
  CONST 0))
# Instructions:
L.4:	# <- 
	call ArrayIndexVariable	# %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 <- 
# Control Flow Graph:
0: %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 <- %rsp %rbp ; goto 
# Interference Graph:
%rax: 
%rbx: 
%rcx: 
%rdx: 
%rsp: 
%rbp: 
%rsi: 
%rdi: 
%r8: 
%r9: 
%r10: 
%r11: 
%r12: 
%r13: 
%r14: 
%r15: 
# Register Allocation:
%rax->%rax
%rbx->%rbx
%rcx->%rcx
%rdx->%rdx
%rsp->%rsp
%rbp->%rbp
%rsi->%rsi
%rdi->%rdi
%r8->%r8
%r9->%r9
%r10->%r10
%r11->%r11
%r12->%r12
%r13->%r13
%r14->%r14
%r15->%r15
# Assembly code:
	.globl main
	.text
main:
	pushq %rbp
	movq %rsp,%rbp
L.4:
	call ArrayIndexVariable
#	returnSink
	popq %rbp
	ret
END main
	.data
	.balign 8
L.1:	.asciz	"Attempt to use a null pointer"
PROCEDURE badPtr
# Before canonicalization (trees):
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.1)),
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# After canonicalization (trees):
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.1))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# Basic Blocks:
LABEL L.5
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.1))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# Trace Scheduled:
LABEL L.5
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.1))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# With procedure entry/exit:
LABEL L.5
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.1))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# Instructions:
L.5:	# <- 
	leaq L.1(%rip),t.4	# t.4 <- 
	movq t.4,%rdi	# %rdi := t.4 
	call puts	# %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 <- %rdi 
	movabsq $1,t.5	# t.5 <- 
	movq t.5,%rdi	# %rdi := t.5 
	call exit	# %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 <- %rdi 
# Control Flow Graph:
0: t.4 %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 t.5 <- %rsp %rbp ; goto 
# Interference Graph:
%rax: 
%rbx: 
%rcx: 
%rdx: 
%rsp: 
%rbp: 
%rsi: 
%rdi: 
%r8: 
%r9: 
%r10: 
%r11: 
%r12: 
%r13: 
%r14: 
%r15: 
t.4: %rsp %rbp 
t.5: %rsp %rbp 
%rdi <= t.4
%rdi <= t.5
# Register Allocation:
%rax->%rax
%rbx->%rbx
%rcx->%rcx
%rdx->%rdx
%rsp->%rsp
%rbp->%rbp
%rsi->%rsi
%rdi->%rdi
%r8->%r8
%r9->%r9
%r10->%r10
%r11->%r11
%r12->%r12
%r13->%r13
%r14->%r14
%r15->%r15
t.4->%rdi
t.5->%rdi
# Assembly code:
	.text
badPtr:
	pushq %rbp
	movq %rsp,%rbp
L.5:
	leaq L.1(%rip),%rdi
#	movq %rdi,%rdi
	call puts
	movabsq $1,%rdi
#	movq %rdi,%rdi
	call exit
#	returnSink
	popq %rbp
	ret
END badPtr
	.data
	.balign 8
L.2:	.asciz	"Subscript out of bounds"
PROCEDURE badSub
# Before canonicalization (trees):
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.2)),
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# After canonicalization (trees):
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.2))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# Basic Blocks:
LABEL L.6
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.2))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# Trace Scheduled:
LABEL L.6
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.2))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# With procedure entry/exit:
LABEL L.6
EXP(
 CALL(
  NAME puts,
  CONST 0,
  NAME L.2))
EXP(
 CALL(
  NAME exit,
  CONST 0,
  CONST 1))
# Instructions:
L.6:	# <- 
	leaq L.2(%rip),t.6	# t.6 <- 
	movq t.6,%rdi	# %rdi := t.6 
	call puts	# %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 <- %rdi 
	movabsq $1,t.7	# t.7 <- 
	movq t.7,%rdi	# %rdi := t.7 
	call exit	# %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 <- %rdi 
# Control Flow Graph:
0: t.6 %rdi %rsi %rdx %rcx %r8 %r9 %rax %r10 %r11 t.7 <- %rsp %rbp ; goto 
# Interference Graph:
%rax: 
%rbx: 
%rcx: 
%rdx: 
%rsp: 
%rbp: 
%rsi: 
%rdi: 
%r8: 
%r9: 
%r10: 
%r11: 
%r12: 
%r13: 
%r14: 
%r15: 
t.6: %rsp %rbp 
t.7: %rsp %rbp 
%rdi <= t.6
%rdi <= t.7
# Register Allocation:
%rax->%rax
%rbx->%rbx
%rcx->%rcx
%rdx->%rdx
%rsp->%rsp
%rbp->%rbp
%rsi->%rsi
%rdi->%rdi
%r8->%r8
%r9->%r9
%r10->%r10
%r11->%r11
%r12->%r12
%r13->%r13
%r14->%r14
%r15->%r15
t.6->%rdi
t.7->%rdi
# Assembly code:
	.text
badSub:
	pushq %rbp
	movq %rsp,%rbp
L.6:
	leaq L.2(%rip),%rdi
#	movq %rdi,%rdi
	call puts
	movabsq $1,%rdi
#	movq %rdi,%rdi
	call exit
#	returnSink
	popq %rbp
	ret
END badSub
